src/cdomains/valueDomain.ml

```diff
@@ -986,13 +986,20 @@ struct
               let new_array_value = CArrays.set ask x' (e, idx) new_value_at_index in
               `Array new_array_value
             | `Bot ->
-              let t = (match t with
-              | TArray(t1 ,_,_) -> t1
-              | _ -> t) in (* This is necessary because t is not a TArray in case of calloc *)
+              let t,len = (match t with
+                  | TArray(t1 ,len,_) -> t1, len
+                  | _ -> t, None) in (* This is necessary because t is not a TArray in case of calloc *)
               let x' = CArrays.bot () in
               let e = determine_offset ask l o exp (Some v) in
               let new_value_at_index = do_update_offset ask `Bot offs value exp l' o' v t in
               let new_array_value =  CArrays.set ask x' (e, idx) new_value_at_index in
+              let newl = match len with
+                | None -> ID.top_of (Cilfacade.ptrdiff_ikind ()) (* TODO: must be non-negative, top is overly cautious *)
+                | Some e ->
+                  let l = BatOption.map (fun x -> IndexDomain.of_int (Cilfacade.ptrdiff_ikind ()) @@ Cilint.big_int_of_cilint x) (Cil.getInteger @@ Cil.constFold true e) in
+                  BatOption.default (ID.top_of (Cilfacade.ptrdiff_ikind ())) l
```

Comment:
I suppose you could just apply `Option.map` on `len` first and then `Option.default`. That way you could avoid duplicating the default value in this branch and the `None` branch above.

Reply:
It has to be bind instead of map since `Cil.getInteger` returns an `option` itself. But apart from that, I modified it accordingly.